File: dotnet/Logify.Alert.Core/Logger/JsonTextWriterLogger.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using DevExpress.Office.Utils;

namespace DevExpress.Logify.Core.Internal {
    public class TextWriterLogger : ILogger {
        #region JsonLoggerObject
        class JsonLoggerObject {
            #region static
            static readonly Dictionary<char, string> replaceTable = CreateReplaceTable();
            static Dictionary<char, string> CreateReplaceTable() {
                Dictionary<char, string> result = new Dictionary<char, string>();
                result.Add('\r', @"\r");
                result.Add('\n', @"\n");
                result.Add('\t', @"\t");
                result.Add('\"', @"\""");
                result.Add('\\', @"\\");
                result.Add('\b', @"\b");
                result.Add('\f', @"\f");
                for (int i = 0; i < 32; i++) {
                    char ch = (char)i;
                    if (!result.ContainsKey(ch))
                        result.Add(ch, String.Format(@"\u{0:x4}", i));
                }
                return result;
            }
            static string replaceDot(string name) {
                //MongoDB doesn't support keys with a dot in them
                return name.Replace('.', '\uff0E');
            }
            #endregion
            #region fields
            string name;
            string value;
            List<JsonLoggerObject> objects;
            readonly bool isArray;
            readonly FastCharacterMultiReplacement replacement = new FastCharacterMultiReplacement(new StringBuilder());
            readonly JsonLoggerObject parent;
            #endregion
            #region Properties
            public JsonLoggerObject Parent { get { return parent; } }
            public bool NeedQuotes { get; set; }
            #endregion
            public JsonLoggerObject() : this(string.Empty, null, null) {
            }
            public JsonLoggerObject(string name, string value, JsonLoggerObject parent) : this(name, value, parent, false) {
            }
            public JsonLoggerObject(string name, string value, JsonLoggerObject parent, bool isArray) {
                this.NeedQuotes = false;
                this.name = name;
                this.value = value;
                this.parent = parent;
                this.isArray = isArray;
            }
            public void Add(JsonLoggerObject jsonLoggerObject) {
                if (this.objects == null) {
                    this.objects = new List<JsonLoggerObject>();
                }
                this.objects.Add(jsonLoggerObject);
            }
            public void SaveContent(TextWriter writer, bool lastObject) {
                if (writer != null) {
                    string jsonName = String.IsNullOrEmpty(name) ? string.Empty : "\"" + replaceDot(name) + "\":";
                    if (objects == null) {
                        string text = !NeedQuotes ? value : replacement.PerformReplacements(value, replaceTable);
                        string quote = !NeedQuotes ? string.Empty : "\"";
                        string comma = lastObject ? string.Empty : ",";
                        if (value == null) {
                            text = isArray ? "[]" : "";
                        }
                        writer.WriteLine(jsonName + quote + text + quote + comma);
                    } else {
                        string openedSymbol = isArray ? "[" : "{";
                        string closedSymbol = isArray ? "]" : "}";
                        writer.WriteLine(jsonName + openedSymbol);
                        if (objects != null) {
                            for (int i = 0; i < objects.Count; i++) {
                                objects[i].SaveContent(writer, i == objects.Count - 1);
                            }
                        }
                        writer.WriteLine(closedSymbol + (lastObject ? string.Empty : ","));
                    }
                }

            }
        }
        #endregion
        #region fields
        readonly TextWriter writer;
        readonly Dictionary<string, object> data = new Dictionary<string, object>();
        JsonLoggerObject lastOpenedObject;
        #endregion
        #region Properties
        public TextWriter Writer { get { return writer; } }
        public Dictionary<string, object> Data { get { return data; } }
        #endregion

        public TextWriterLogger(TextWriter writer) {
            this.writer = writer;
            lastOpenedObject = null;
        }

        public void BeginWriteObject(string name) {
            BeginWriteObject(name, false);
        }
        public void EndWriteObject(string name) {
            EndWriteObject();
        }
        public void BeginWriteArray(string name) {
            BeginWriteObject(name, true);
        }
        public void EndWriteArray(string name) {
            EndWriteObject();
        }
        public void WriteValue(string name, string text) {
            JsonLoggerObject jsonLoggerObject = new JsonLoggerObject(name, text, lastOpenedObject);
            jsonLoggerObject.NeedQuotes = true;
            IncludeInObjectModel(jsonLoggerObject);
        }
        public void WriteValue(string name, bool value) {
            IncludeInObjectModel(new JsonLoggerObject(name, (value ? "true" : "false"), lastOpenedObject));
        }
        public void WriteValue(string name, int value) {
            IncludeInObjectModel(new JsonLoggerObject(name, value.ToString(), lastOpenedObject));
        }
        public void WriteValue(string name, Array array) {
            if (array == null || array.Length == 0) return;
            IncludeInObjectModel(new JsonLoggerObject(name, ArrayToString(array), lastOpenedObject));
        }

        void BeginWriteObject(string name, bool isArray) {
            JsonLoggerObject newObject = new JsonLoggerObject(name, null, lastOpenedObject, isArray);
            IncludeInObjectModel(newObject);
            lastOpenedObject = newObject;
        }
        void EndWriteObject() {
            if (lastOpenedObject == null)
                return;
            if (lastOpenedObject.Parent != null) {
                lastOpenedObject = lastOpenedObject.Parent;
            } else {
                lastOpenedObject.SaveContent(writer, true);
                lastOpenedObject = null;
            }
        }
        string ArrayToString(Array array) {
            string result = "[";
            System.Collections.IEnumerator iterator = array.GetEnumerator();
            for (int i = 0; iterator.MoveNext(); i++) {
                result += "\"" + iterator.Current.ToString() + "\"";
                if (i != array.Length - 1) result += ",";
            }
            result += "]";
            return result;
        }
        void IncludeInObjectModel(JsonLoggerObject jsonLoggerObject) {
            if (lastOpenedObject != null)
                lastOpenedObject.Add(jsonLoggerObject);
            else
                jsonLoggerObject.SaveContent(writer, true);
        }
    }
    
}

File: java/Core/src/main/java/com/devexpress/logify/alert/core/logger/JsonTextWriterLogger.java
package com.devexpress.logify.alert.core.logger;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class JsonTextWriterLogger implements ILogger {

    private final Map<String, Object> data = new HashMap<String, Object>();
    private final StringBuilder builder;
    private boolean isFirstElement = true;

    public JsonTextWriterLogger(StringBuilder builder) {
        this.builder = builder;
    }

    public StringBuilder getBuilder() {
        return builder;
    }

    public Map<String, Object> getData() {
        return data;
    }

    private static String escapeText(String text) {
        int index;
        for (index = 0; index < text.length(); index++) {
            char c = text.charAt(index);
            if (c < 0x20 || c == '\"' || c == '\\')
                break;
        }

        if (index == text.length())
            return text;

        StringBuilder result = new StringBuilder();
        result.append(text, 0, index);
        for (int i = index; i < text.length(); i++) {
            char c = text.charAt(i);
            if (c == '\"') {
                result.append("\\\"");
                continue;
            }
            if (c == '\\') {
                result.append("\\\\");
                continue;
            }
            if (c >= 0x20) {
                result.append(c);
                continue;
            }
            result.append(String.format("\\u%04x", (int) c));
        }
        return result.toString();
    }

    private void writeName(String name) {
        builder.append("\"")
                .append(replaceDot(name))
                .append("\":");
    }

    public void beginWriteObject(String name) {
        if (!isFirstElement)
            builder.append(",");
        isFirstElement = true;

        if (!name.isEmpty())
            writeName(name);
        builder.append("{");
    }

    public void writeValue(String name, String text) {
        if (!isFirstElement)
            builder.append(",");
        isFirstElement = false;

        writeName(name);
        if(text != null)
            builder.append("\"")
                .append(escapeText(text))
                .append("\"");
        else
            builder.append("null");
    }

    public void writeValue(String name, boolean value) {
        if (!isFirstElement)
            builder.append(",");
        isFirstElement = false;

        writeName(name);
        builder.append(value ? "true" : "false");
    }

    public void writeValue(String name, int value) {
        if (!isFirstElement)
            builder.append(",");
        isFirstElement = false;

        writeName(name);
        builder.append(value);
    }

    public void writeValue(String name, List array) {
        beginWriteArray(name);
        for (int i = 0; i < array.size(); i++) {
            builder.append("\"").append(array.get(i).toString()).append("\"");
            if (i != array.size() - 1)
                builder.append(",");
        }
        endWriteArray(name);
    }

    public void endWriteObject(String name) {
        builder.append("}");

        isFirstElement = false;
    }

    public void beginWriteArray(String name) {
        if (!isFirstElement)
            builder.append(",");
        isFirstElement = true;

        writeName(name);
        builder.append("[");
    }

    public void endWriteArray(String name) {
        builder.append("]");

        isFirstElement = false;
    }

    private String replaceDot(String name) {
        //MongoDB doesn't support keys with a dot in them
        return name.replace('.', '\uff0E');
    }

    @Override
    public String toString() {
        return builder.toString();
    }
}